### PR TITLE
Adds a function to get linearly spaced astropy time object in TimeRange

### LINKED ISSUE #5377: Functionality to pass a dt (sampling frequency) to TimeRange and it would return list of astropy time objects
### Description

It would be nice if you could pass a start and end time and a sampling frequency to TimeRange and it would return a list of astropy time objects. 

The list could be an attribute of the TimeRange class

E.g. an equivalent to something like np.linspace but for a timerange

### PR DESCRIPTION
**Pull request by Krish2208**
## PR Description

Adds a function `range` to TimeRange Class which returns a linearly spaced astropy time object, as mentioned in the issue #5377 

Fixes #5377 

## PR Checklist

- [x] I have followed the guidelines in the [Contributing document](https://docs.sunpy.org/en/latest/dev_guide/contents/newcomers.html)
- [x] Changes follow the coding style of this project
- [x] Changes have been formatted and linted
- [ ] Changes pass pytest style unit tests (and `pytest` passes).
- [ ] Changes include any required corresponding changes to the documentation
- [x] Changes have been rebased and squashed to the minimal number (typically 1) of relevant commits
- [x] Changes have a descriptive commit message with a short title
- [ ] I have added a `Fixes #XXXX -` or `Closes #XXXX -` comment to auto-close the issue that your PR addresses

### CODE REVIEW

**Comment by nabobalis** [comment]
Can I get an example of how someone uses this?

**Comment by Krish2208** [comment]
> Can I get an example of how someone uses this?



```
tr = TimeRange('2022/03/31 12:00', '2022/03/31 13:00')
range_list = tr.range_list(5)
print(range_list)
```

This would print a list of 5 linearly spaced astropy time objects

`[<Time object: scale='utc' format='isot' value=2022-03-31T12:00:00.000>, <Time object: scale='utc' format='isot' value=2022 03-31T12:15:00.000>, <Time object: scale='utc' format='isot' value=2022-03-31T12:30:00.000>, <Time object: scale='utc' format='isot' value=2022-03-31T12:45:00.000>, <Time object: scale='utc' format='isot' value=2022-03-31T13:00:00.000>]`

File: sunpy/time/timerange.py [hunk 1]
```diff
@@ -460,3 +461,20 @@ def intersects(self, other):
             int_second = self
 
         return int_second.start <= int_first.end
+
+    def range_list(self, frequency):
```
Comment (nabobalis) [inline sunpy/time/timerange.py:465]: Could we not name this range?

File: sunpy/time/timerange.py [hunk 2]
```diff
@@ -460,3 +461,20 @@ def intersects(self, other):
...
+        """
+        Return a Linearly Spaced List of Astropy Time Objects between the passed TimeRange.
```
Comment (nabobalis) [inline sunpy/time/timerange.py:467]:
```suggestion
        Return a linearly spaced time range.
```

File: sunpy/time/timerange.py [hunk 3]
```diff
@@ -460,3 +461,20 @@ def intersects(self, other):
...
+
+        Parameters
+        ----------
+        frequency : `int`
+            The number of objects needed
```
Comment (nabobalis) [inline sunpy/time/timerange.py:472]:
```suggestion
        steps : `int`
            The number of steps between the start and end time.
```

File: sunpy/time/timerange.py [hunk 4]
```diff
@@ -460,3 +461,20 @@ def intersects(self, other):
...
+        """
+        if frequency<=0:
+            raise ValueError("The number of objects needed cannot be 0 or less")
```
Comment (nabobalis) [inline sunpy/time/timerange.py:475]: I am not sure if we need this check?

File: sunpy/time/timerange.py [hunk 5]
```diff
@@ -460,3 +461,20 @@ def intersects(self, other):
...
+        end = self._t2
+        start = self._t1
+        delta = (end-start)
```
Comment (nabobalis) [inline sunpy/time/timerange.py:478]: I don't think we need these defined as variables?

File: sunpy/time/timerange.py [hunk 6]
```diff
@@ -460,3 +461,20 @@ def intersects(self, other):
...
+        time_list = list(start + delta * np.linspace(0,1,frequency))
```
Comment (nabobalis) [inline sunpy/time/timerange.py:479]: It should be possible to return one astropy.time.Time object that has a range of values, I would personally prefer that instead of a list.

**Comment by nabobalis** [comment]
> > Can I get an example of how someone uses this?
> 
> ```
> tr = TimeRange('2022/03/31 12:00', '2022/03/31 13:00')
> range_list = tr.range_list(5)
> print(range_list)
> ```
> 
> This would print a list of 5 linearly spaced astropy time objects
> 
> `[<Time object: scale='utc' format='isot' value=2022-03-31T12:00:00.000>, <Time object: scale='utc' format='isot' value=2022 03-31T12:15:00.000>, <Time object: scale='utc' format='isot' value=2022-03-31T12:30:00.000>, <Time object: scale='utc' format='isot' value=2022-03-31T12:45:00.000>, <Time object: scale='utc' format='isot' value=2022-03-31T13:00:00.000>]`

Thanks!

**Comment by nabobalis** [comment]
Just now to add some unit tests and a changelog entry!

File: changelog/6017.feature.rst [hunk 7]
```diff
@@ -0,0 +1 @@
+Added a range function to TimeRange class that returns a linearly spaced time object.
```
Comment (nabobalis) [inline changelog/6017.feature.rst:1]:
```suggestion
Added a range function to 	`.TimeRange` that returns a linearly spaced time object.
```

File: sunpy/time/tests/test_timerange.py [hunk 8]
```diff
@@ -292,3 +292,10 @@ def test_intersects():
     tr2 = sunpy.time.TimeRange('2020-01-01', '2020-01-04')
     assert tr1.intersects(tr2)
     assert tr2.intersects(tr1)
+
+def test_range():
```
Comment (nabobalis) [inline sunpy/time/tests/test_timerange.py:296]: Can I get a few more tests, with larger ranges, what happens if you pass in a negative one or 0 even.

**Comment by nabobalis** [comment]
I looked over the original issue, and the request was for a sampling frequency. 

I wonder if passing a astropy.time.TimeDelta would be a better idea than a int?

**Comment by Krish2208** [comment]
> I looked over the original issue, and the request was for a sampling frequency.
> 
> I wonder if passing a astropy.time.TimeDelta would be a better idea than a int?

So we need a something like 
When we pass `TimeDelta(300, format="sec")` into the function, we should get a time object with 5 min difference between consecutive values?

**Comment by nabobalis** [comment]
> So we need a something like When we pass `TimeDelta(300, format="sec")` into the function, we should get a time object with 5 min difference between consecutive values?

Yeah,  that sounds right.
 
> Shall I add another method for this or do I need to make changes to the `range()` method?

I think it should be a change to the range function.

File: changelog/6017.feature.rst [hunk 9]
```diff
@@ -0,0 +1 @@
+Added a range function to `TimeRange` that returns a linearly spaced time object.
```
Comment (nabobalis) [inline changelog/6017.feature.rst:1]:
```suggestion
Added a range function to `.TimeRange` that returns a linearly spaced time object.
```

**Review by dstansby — DISMISSED** [review]
Thanks for working on this; I'm personally not sure we should implement this ourselves, but instead direct users using `np.arange(TimeRange.start, TimeRange.end, delta)`. This:
- Avoids us creating new API
- Is a more general solution, that users can use with any two times

As such I'm going to block this for now until there's been a wider developer discussion. I'd say if someone else agrees with me we should close this without merging (sorry), and turn https://github.com/sunpy/sunpy/issues/5377 into a request to add `np.arange(TimeRange.start, TimeRange.end, delta)` to the docs somewhere.

I'm willing to be swayed on my thoughts though if other developers think we should add this?

**Comment by nabobalis** [comment]
I have to agree with dstansby, having thought about it and shown that anyone can do this with `np.arange(TimeRange.start, TimeRange.end, delta)`, I think maybe a mention in the documentation of time range as an example would be a better idea than a new function.

We can turn this PR into that?

**Comment by Krish2208** [comment]
I added an example can you please check it.
Sorry for the delay.

File: sunpy/time/timerange.py [hunk 10]
```diff
@@ -65,6 +65,16 @@ class TimeRange:
     True
     >>> time2 in time_range
     False
+    >>> import numpy as np
```
Comment (nabobalis) [inline sunpy/time/timerange.py:68]:
```suggestion

    >>> import numpy as np
```

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
Thanks for the PR @Krish2208
